lib/ret/media_resolver.ex

```diff
@@ -34,7 +34,26 @@ defmodule Ret.MediaResolver do
   def resolve(%MediaResolverQuery{url: url} = query) when is_binary(url) do
     uri = url |> URI.parse()
     root_host = get_root_host(uri.host)
-    resolve(query |> Map.put(:url, uri), root_host)
+    query = Map.put(query, :url, uri)
+
+    # TODO: We could end up running fallback_to_screenshot_opengraph_or_nothing
+    #       twice in a row. These resolve functions can be simplified so that we can
+    #       more easily track individual failures and only fallback when necessary.
+    #       Also make sure they have a uniform response shape for indicating an
+    #       error.
```

Comment:
This is a tech debt concern that I do not want to cover in this PR, (and I am not sure I'm right), but it might be worth changing in the future:

I think what we might want to do is:
- Refactor `resolve/2`, `resolve_with_ytdl`, and `resolve_non_video` so that they return `{:ok, %ResolvedMedia{}} | {:error, reason}` instead of a return value for the `Cachex` `fallback` function.
- In the outer function (`resolve/1`), match on the `{:ok, %ResolvedMedia{}} | {:error, reason}` tuples (from the functions above) using `case` to decide 
  - whether to try to fallback to screenshots/opengraph,
  - whether to increment Statix metrics or other telemetry,
  - what we _actually_ want to store in the cache in the event of an error (e.g. details about what went wrong so that we can return that info to the client)
- From the outer function (`resolve/1`), return a tuple for the `Cachex` `fallback` function with a status of `:commit` or `:ignore`:
> A fallback function should return a Tuple consisting of a :commit or :ignore tag and a value. If the Tuple is tagged :commit the value will be placed into the cache and then returned. If tagged :ignore the value will be returned without being written to the cache. If you return a value which does not fit this structure, it will be assumed that you are committing the value.

From https://hexdocs.pm/cachex/Cachex.html#fetch/4

This may not be necessary, but it seems to me like it might be clearer to explicitly return one of:
  - `{:commit, resolved_media}`
  - `{:commit, {:error, reason}}`
  - ~~`{:commit, nil}`~~ (Don't use `nil` to indicate 404 -- instead use something like `{:commit, {:error, 404}}`
  - `{:ignore, some_uncached_value}`
  
  We should do our best differentiate between the return values of `Cachex.fetch(:media_urls, query)` responses. Specifically, 
  ``` elixir
  case Cachex.fetch(:media_urls, query) do
    {:error, _} -> # This is a Cachex error!
    {_status, {:error, _reason} } -> # This is a cached error we put in ourselves
  end
  ```
  See this commit for details https://github.com/mozilla/reticulum/pull/452/commits/ffab45ce844dcd6f0aac6e86a4b93c37e966c244#diff-41751f7931f9a08bf162d96c927b4c6872c21b2d215916a5b0f8041848d08152R119
  
This might be easier to distinguish if we create a `ResolvedMediaError` struct to put in the cache so that we don't mistake the difference between the two.

Reply:
This sounds correct to me. Mixing the fallback stuff in with resolving is confusing. We should have resolving return an ok/error result like you suggest, and then use that in cachex fallback. Might be nice for the cachex part not even to live in this file.

Reply:
Captured this in https://github.com/mozilla/reticulum/issues/454 for addressing at a later date